node/silkworm/downloader/sentry_client.cpp

```diff
@@ -79,9 +88,36 @@ void SentryClient::execution_loop() {
         publish(message);
     }
 
-    // note: do we need to handle connection loss retrying re-connect? (we would redo set_status too)
+    // note: do we need to handle connection loss with an outer loop that wait and than re-try hand_shake and so on? (we would redo set_status too)
 
     log::Warning() << "SentryClient execution loop is stopping...";
 }
 
+void SentryClient::stats_receiving_loop() {
+    // send a stats subscription
+    rpc::ReceivePeerStats receive_peer_stats;
+    exec_remotely(receive_peer_stats);
+
+    // receive stats
+    int active_peers = 0;
+    while (!is_stopping() && receive_peer_stats.receive_one_reply()) {
+        const sentry::PeersReply& stat = receive_peer_stats.reply();
+
+        auto peerId = string_from_H512(stat.peer_id());
+        const char* event = "";
+        if (stat.event() == sentry::PeersReply::Connect) {
+            event = "connected";
+            active_peers++;
+        }
+        else {
+            event = "disconnected";
+            active_peers--;
+        }
+
+        log::Trace() << "SentryClient: peer " << peerId << " " << event << ", active " << active_peers;
+    }
+
+    log::Warning() << "SentryClient stats loop is stopping...";
```

Comment:
If I read this I understand the while loop exits if `is_stopping` or `receive_peer_stats.receive_one_reply()` returns false
So why the log::Warning reports only the client is stopping ? Isn't possible we simply don't have answers yet ?

Reply:
`receive_peer_stats` is a stream of messages (replies in gRPC naming convention) from the server; the RPC call `receive_peer_stats.receive_one_reply()` is a blocking one and wait for the next reply.

Sure the code does not handle connection loss with the sentry. I avoided to repeat the comment that you can read in the method preceding this one:  "_do we need to handle connection loss with an outer loop that wait and than re-try hand_shake and so on? (we would redo set_status too)_"